### PR TITLE
Add exception handler for model server and Add ability to specify custom handler

### LINKED ISSUE #3404: Add exception handler for model server
/kind bug
Model server shutdown is not properly handled when an Exception occurs.
```
Traceback (most recent call last):
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/sklearnserver/sklearnserver/__main__.py", line 41, in <module>
    kserve.ModelServer(registered_models=SKLearnModelRepository(args.model_dir), workers=6).start(
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/kserve/kserve/model_server.py", line 229, in start
    asyncio.run(servers_task())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/kserve/kserve/model_server.py", line 227, in servers_task
    await asyncio.gather(*servers)
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/kserve/kserve/model_server.py", line 214, in serve
    raise RuntimeError("Some Exception")
RuntimeError: Some Exception
Exception ignored in: <function Server.__del__ at 0x7f90add16170>
Traceback (most recent call last):
  File "/home/ubuntu/go/src/github.com/kserve/kserve/.venv/lib/python3.10/site-packages/grpc/aio/_server.py", line 185, in __del__
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 118, in grpc._cython.cygrpc.schedule_coro_threadsafe
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 110, in grpc._cython.cygrpc.schedule_coro_threadsafe
  File "/usr/lib/python3.10/asyncio/base_events.py", line 436, in create_task
  File "/usr/lib/python3.10/asyncio/base_events.py", line 515, in _check_closed
RuntimeError: Event loop is closed
sys:1: RuntimeWarning: coroutine 'AioServer.shutdown' was never awaited
```
**Solution**:
Add an exception handler to the event loop to handle the shutdown process.

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

**What did you expect to happen:**
The model server should be properly shutdown when an exception occurs.

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version: 0.10 - 0.12.0-rc1
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:
- Handles Model server shutdown properly when exception occured.
- Provides a way to add custom exception handler to the event loop

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3404

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [x] Test A
Before:
```
Traceback (most recent call last):
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/sklearnserver/sklearnserver/__main__.py", line 41, in <module>
    kserve.ModelServer(registered_models=SKLearnModelRepository(args.model_dir), workers=6).start(
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/kserve/kserve/model_server.py", line 229, in start
    asyncio.run(servers_task())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/kserve/kserve/model_server.py", line 227, in servers_task
    await asyncio.gather(*servers)
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/kserve/kserve/model_server.py", line 214, in serve
    raise RuntimeError("Some Exception")
RuntimeError: Some Exception
Exception ignored in: <function Server.__del__ at 0x7f90add16170>
Traceback (most recent call last):
  File "/home/ubuntu/go/src/github.com/kserve/kserve/.venv/lib/python3.10/site-packages/grpc/aio/_server.py", line 185, in __del__
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 118, in grpc._cython.cygrpc.schedule_coro_threadsafe
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 110, in grpc._cython.cygrpc.schedule_coro_threadsafe
  File "/usr/lib/python3.10/asyncio/base_events.py", line 436, in create_task
  File "/usr/lib/python3.10/asyncio/base_events.py", line 515, in _check_closed
RuntimeError: Event loop is closed
sys:1: RuntimeWarning: coroutine 'AioServer.shutdown' was never awaited
```

After:
```
Traceback (most recent call last):
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/sklearnserver/sklearnserver/__main__.py", line 41, in <module>
    kserve.ModelServer(registered_models=SKLearnModelRepository(args.model_dir), workers=6).start(
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/kserve/kserve/model_server.py", line 224, in start
    asyncio.run(servers_task())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/kserve/kserve/model_server.py", line 222, in servers_task
    await asyncio.gather(*servers)
  File "/home/ubuntu/go/src/github.com/kserve/kserve/python/kserve/kserve/model_server.py", line 209, in serve
    raise RuntimeError("Some Exception")
RuntimeError: Some Exception
```

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Add exception handler for model server and Add ability to specify custom exception handler
```

### CODE REVIEW

File: python/kserve/kserve/model_server.py [hunk 1]
```diff
@@ -231,6 +233,27 @@ async def stop(self, sig: Optional[int] = None):
             logger.info("Stopping the grpc server")
             await self._grpc_server.stop(sig)
 
+    def add_exception_handler(self, func: Callable[[asyncio.events.AbstractEventLoop, Dict[str, Any]], None]):
```
Comment (terrytangyuan) [inline python/kserve/kserve/model_server.py]: Consider renaming it to `register_exception_handler` for consistency?

**Comment by rachitchauhan43** [comment]
@sivanantha321 : Possible to get a unit test case around this feature to test that custom exception handler is indeed getting added ?

**Review by rachitchauhan43 — COMMENTED** [review]
Left a comment regarding unit test , rest LGTM

File: python/kserve/kserve/model_server.py (hunk 1 above)
Comment (sivanantha321) [reply python/kserve/kserve/model_server.py]: Renamed

**Comment by sivanantha321** [comment]
> @sivanantha321 : Possible to get a unit test case around this feature to test that custom exception handler is indeed getting added ?

I tried. But I am not able to get it to work.

**Review by rachitchauhan43 — APPROVED** [review]

**Pull request merged**